4.next - Added IntervalExpression to handle datetime interval addition/subtraction in the orm

jimbo2150 opened the pull request:
This PR adds support for datetime intervals without having to use a workaround and will translate the syntax across different servers.

The MySQL syntax is `dateOrField + INTERVAL 2 YEAR + INTERVAL 3 DAY`
The PostgrSQL syntax is `dateOrField + INTERVAL '2 YEAR 3 DAY'`
The SQLite syntax is `strftime('%Y-%m-%d %H:%M:%f', dateOrField, '2 year', '3 day')`
- The strfttime syntax is for the use of fractional time. Other functions such as `date()` do not support fractions of a second. The result can then be passed through date/time conversion to get the desired output.

The MSSQL syntax is `date_add('DAYS', 3, date_add('YEARS', 2, dateOrField))`

Example usage:
```php
$interval = \DateInterval::createFromDateString('2 year + 3 day');
$iExp = new IntervalExpression(new FrozenTime('2020-04-17 02:03:02.21'), $interval);
(new Query($connection))->select([ $iExp ]);
```
Would produce output like
```sql
SELECT '2020-04-17 02:03:02.21' + INTERVAL 2 YEAR + INTERVAL 3 DAY;
```
in MySQL and should produce a result `2022-04-20 02:03:02.21`.

The requiring of a date or expression was necessary for supporting other databases since SQLite and MSSQL do not support INVERVAL expressions, only date modifier functions which require an expression or date to be passed as a parameter.

Special and relative intervals such as "Second Thursday in October" are not supported because there is no meaningful way to acquire that interval from the PHP interval object and most databases do not support relative/special intervals.

jimbo2150 committed:
- Added IntervalExpression to handle date interval addition/subtraction in the orm.

jimbo2150 committed:
- Corrected plural suffix being added to value rather than unit

dereuromark commented on src/Database/Expression/IntervalExpression.php:
```diff
@@ -0,0 +1,309 @@
+    /**
+     * SQL unit values for DateInterval array conversion.
+     *
+     * @var array
```
string[]

jimbo2150 replied:
Corrected

jimbo2150 committed:
- Corrected si errors

jimbo2150 committed:
- Corrected comment for class

dereuromark commented on src/Database/Expression/IntervalExpression.php:
```diff
@@ -0,0 +1,309 @@
+    public function setInterval(DateInterval $interval): self
+    {
+        $isValid = ($interval->y + $interval->m + $interval->d +
+            $interval->h + $interval->i + $interval->s + $interval->f) != 0;
```
!==

jimbo2150 replied:
Corrected

dereuromark commented on src/Database/Expression/IntervalExpression.php:
```diff
@@ -0,0 +1,309 @@
+    {
+        $isValid = ($interval->y + $interval->m + $interval->d +
+            $interval->h + $interval->i + $interval->s + $interval->f) != 0;
+        if ($isValid == false) {
```
if (!$isValid)

jimbo2150 committed:
- Fixed si errors

jimbo2150 committed:
- Added strict type check
- Corrected property identifier

jimbo2150 committed:
- Removed debug block

jimbo2150 committed:
- Corrected prefix and suffix placement

jimbo2150 committed:
- Corrected incorrect date modification function name for MSSQL

jimbo2150 committed:
- Additional lint corrections

jimbo2150 committed:
- Corrected lint and suggested changes
- Fixed additional comma in fucntion call

othercorey commented:
Is the goal to use this with WindowExpressions or just adding for general use?

jimbo2150 commented:
> Is the goal to use this with WindowExpressions or just adding for general use?

Both. Though WindowExpressions will be a second iteration since use with a date requires a value to be added due to not all servers supporting the interval expression.

jimbo2150 committed:
- Updated to work with PostgreSQL's requirement of a cast on dates

jimbo2150 committed:
- Updated options so that dates are properly cast to timestamp

jimbo2150 committed:
- Added optional separation of microseconds (SQL Server requires it)

jimbo2150 committed:
- Keeping microseconds separate (required by SQL Server)
- Removed plural units

jimbo2150 committed:
- Removed debug

markstory commented on src/Database/Driver/Postgres.php:
```diff
@@ -297,4 +299,22 @@ public function newCompiler(): QueryCompiler
+            'glue' => ' ',
+            'sql-prefix' => 'INTERVAL ',
+            'prefix' => '\'',
+            'suffix' => '\'',
```
Could use `"'"` to avoid needing to escape quotes.

markstory commented on src/Database/Driver/Sqlite.php:
```diff
@@ -336,4 +338,30 @@ public function supportsWindowFunctions(): bool
+     */
+    protected function transformIntervalExpression(IntervalExpression $intervalExp): void
+    {
+        $fOrV = $intervalExp->getFieldOrValue();
```
Could we use a more readable variable name here? Perhaps `$fieldValue`?

markstory commented on src/Database/Driver/Sqlite.php:
```diff
@@ -336,4 +338,30 @@ public function supportsWindowFunctions(): bool
+    protected function transformIntervalExpression(IntervalExpression $intervalExp): void
+    {
+        $fOrV = $intervalExp->getFieldOrValue();
+        $typeAry = [];
```
Could be `$types` instead.

markstory commented on src/Database/Driver/Sqlite.php:
```diff
@@ -336,4 +338,30 @@ public function supportsWindowFunctions(): bool
+            $typeAry[] = 'datetimefractional';
+        }
+        $sign = $intervalExp->getIntervalSign();
+        $fncExp = (new FunctionExpression('strftime', ['\'%Y-%m-%d %H:%M:%f\'' => 'literal']))
```
Could be `$function` instead of `fncExp`

markstory commented on src/Database/Expression/IntervalExpression.php:
```diff
@@ -0,0 +1,316 @@
+
+    /**
+     * Method that sets an overriding expression for servers that require
+     * the use of en expression (usually a function) instead of an interval
```
```suggestion
     * the use of an expression (usually a function) instead of an interval
```

markstory commented on src/Database/Expression/IntervalExpression.php:
```diff
@@ -0,0 +1,316 @@
+     * @param \DateTimeInterface|\Cake\Database\ExpressionInterface $fov A DateTimeInterface or ExpressionInterface object.
+     * @return self
+     */
+    public function setFieldOrValue($fov): self
```
Could this be 'target' or 'subject'? FieldOrValue is a lot of words to repeat.

markstory reviewed (commented):
Should interval expressions be exposed on the function builder?

I also question the need for all the public methods on `IntervalExpression` it creates a wide public API that doesn't seem to be needed based on what is in the tests. Making those methods protected/private would help us adapt to change more easily in the future.

jimbo2150 committed:
- Added wrapping parentheses

jimbo2150 committed:
- Made various suggested changes

jimbo2150 committed:
- Added subquery test

jimbo2150 committed:
- Replaced single quote with double

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

# Conflicts:
#	src/Database/Driver/Postgres.php
#	src/Database/Driver/Sqlite.php
#	src/Database/Driver/Sqlserver.php

jimbo2150 committed:
- Fixed lint issues

jimbo2150 committed:
- Using static setup and teardown methods

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

dereuromark commented on src/Database/Expression/IntervalExpression.php:
```diff
@@ -0,0 +1,329 @@
+    /**
+     * @inheritDoc
+     */
+    public function traverse(Closure $callable): self
```
No typehint please on return, as this is chainable.

jimbo2150 replied:
@dereuromark These are fixed locally. The editor wasn't warning so I looked and found the setting was still on PS12. Thanks for your help with these!

dereuromark commented on src/Database/Expression/IntervalExpression.php:
```diff
@@ -0,0 +1,329 @@
+     * statement.
+     *
+     * @param \Cake\Database\ExpressionInterface|null $exp An ExpressionInterface object or null value.
+     * @return self
```
$this

dereuromark commented on src/Database/Expression/IntervalExpression.php:
```diff
@@ -0,0 +1,329 @@
+     * @param \Cake\Database\ExpressionInterface|null $exp An ExpressionInterface object or null value.
+     * @return self
+     */
+    public function setOverrideExpression(?ExpressionInterface $exp): self
```
same here, no return typehint allowed for chainable

jimbo2150 commented:
> Should interval expressions be exposed on the function builder?
> 
> I also question the need for all the public methods on `IntervalExpression` it creates a wide public API that doesn't seem to be needed based on what is in the tests. Making those methods protected/private would help us adapt to change more easily in the future.

@markstory, Do you mean within the driver-specific methods? I see your point about having so many exposed functions. I am going to split the expression and move the date code into it's own class since having a date involved is very different from one server to another. It should also make it easier to work with intervals for window expressions.

jimbo2150 committed:
- Applied suggestions

jimbo2150 committed:
Merge branch '4.next' into working/4.1-orm-interval

othercorey commented:
I think this will be easier if use a standard SQL format for intervals.

`DAYS HH:MM:SS.MICROSECOND`

jimbo2150 commented:
> I think this will be easier if use a standard SQL format for intervals.
> 
> `DAYS HH:MM:SS.MICROSECOND`

@othercorey Thanks. I just tested it in pg and it does work even though I don't see that format in the documentation. I could use that to shorten the length of those that include days/hours/minutes/seconds/micro.

jimbo2150 committed:
Merge branch '4.next' into working/4.1-orm-interval

othercorey commented on tests/TestCase/Database/Expression/IntervalExpressionTest.php:
```diff
@@ -0,0 +1,120 @@
+            new FrozenTime(self::$data['date'], self::$data['tz']),
+            self::$data['interval']
+        );
+        $query = new Query(self::$connection);
```
This unit test is difficult to read. Can we follow the format used by other query tests without splitting up every line?

jimbo2150 replied:
I can collapse it into fewer lines. It is split up as I was doing debugging when I needed to see what sql was being produced. Most of the other tests don't actually run queries, however the output of an interval involving a date will be different depending on the server. I found that actually running a query then checking to make sure the result was valid was the best way to test it.

othercorey commented on tests/TestCase/Database/Expression/IntervalExpressionTest.php:
```diff
@@ -0,0 +1,120 @@
+        self::$connection = self::$data = null;
+        parent::tearDownAfterClass();
+    }
+
```
This needs tests that validate the generated sql.

othercorey commented on tests/TestCase/Database/Expression/IntervalExpressionTest.php:
```diff
@@ -0,0 +1,120 @@
+    {
+        // Create temporary table and populate with date
+        $query = new Query(self::$connection);
+        self::$connection->execute('INSERT INTO interval_test VALUES (\'' . self::$data['date'] . '\')');
```
We don't need to do this. Just query an existing table like Comments.

othercorey commented on tests/TestCase/Database/Expression/IntervalExpressionTest.php:
```diff
@@ -0,0 +1,120 @@
+        self::$connection->execute('INSERT INTO interval_test VALUES (\'' . self::$data['date'] . '\')');
+        // Query using subquery
+        $iExp = new IntervalExpression(
+            (new Query(self::$connection))->select(['interval_date'])->from('interval_test')->limit(1),
```
We don't need to test subqueries.

jimbo2150 committed:
- Using closures

jimbo2150 committed:
- Now using options array instead of closure

jimbo2150 committed:
- Completed TSQL

jimbo2150 committed:
Merge branch '4.next' into working/4.1-orm-interval

jimbo2150 committed:
- Updated documentation and corrected sniffer issues

jimbo2150 committed:
- Added exception for unsupported syntax

jimbo2150 committed:
- Switched to using local variable

jimbo2150 committed:
- Switched to using local variable

jimbo2150 committed:
- Updated to using sql options array where possible

jimbo2150 committed:
- Not used import removed

jimbo2150 committed:
- Corrected wrong type hint

jimbo2150 commented:
Any assistance with SQL Server would be helpful as I am not as familiar with it. I ran a bunch of tests locally and the only issue I could find with it complaining about date conversion is if you include a timezone offset. I tried with formatting it as a string without the timezone and got the same error.

jimbo2150 committed:
- Shortened test

jimbo2150 committed:
- Corrected psalm-reported type issue

jimbo2150 commented:
> Any assistance with SQL Server would be helpful as I am not as familiar with it. I ran a bunch of tests locally and the only issue I could find with it complaining about date conversion is if you include a timezone offset. I tried with formatting it as a string without the timezone and got the same error.

So it turns out SQL Server will throw that type error for datetimes including more than 3 fractional second digits unless you explicitly cast it as a `datetime2`.
https://superuser.com/questions/1026511/convert-datetime-with-decimal-time-longer-than-3-digits

jimbo2150 committed:
Merge branch '4.next' into working/4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
- Added option for combining microseconds when transforming interval array

jimbo2150 committed:
- Wrapped date with cast as MS SQL requires an explicit cast to use more than 3 digits within fractional seconds
- Set combineMicro option to false so that seconds and microseconds are separate indexes

jimbo2150 committed:
Merge branch 'working/4.1-orm-interval' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' of https://github.com/cakephp/cakephp into 4.1-orm-interval

jimbo2150 committed:
- Removed unused code

jimbo2150 committed:
- Corrected check for valid interval and added test of invalid relative interval.

jimbo2150 committed:
- Updated subject exception text
- Added test for invalid subject

jimbo2150 committed:
- Moved interval-specific code into a separate object
- Corrected support for mixed sign interval values
- Added test of mixed/negative interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

markstory commented on tests/TestCase/Database/Expression/IntervalExpressionTest.php:
```diff
@@ -0,0 +1,159 @@
+use Cake\Datasource\ConnectionManager;
+use Cake\I18n\FrozenTime;
+use Cake\TestSuite\TestCase;
+use DateInterval as CakeDateInterval;
```
Why is this alias being used?

jimbo2150 replied:
Corrected. Was left over from a previous test though not sure how it got on the base class. It does remind me that I was going to ask and make sure (question posted on the file review you did below).

markstory commented on tests/TestCase/Database/Expression/IntervalExpressionTest.php:
```diff
@@ -0,0 +1,159 @@
+    {
+        parent::setUp();
+        $this->connection = ConnectionManager::get('test');
+        self::$data['tz'] = $utc = new DateTimeZone('UTC');
```
Is there a reason you chose to use class properties instead of instance properties?

jimbo2150 replied:
Left over from when I was testing the static method. Changed to instance property.

markstory commented on src/Database/DateInterval.php:
```diff
@@ -0,0 +1,158 @@
+/**
+ * TODO: Add documentation
+ */
+class DateInterval extends \DateInterval
```
I didn't see this used in the tests anywhere, do we still need it?

jimbo2150 replied:
Added tests and missing documentation.

jimbo2150 replied:
Also wanted to make sure: Do you think this class will be confusing for people? If so, I can rename it. I don't see any classes with the same name as their base and wanted to make sure there wasn't a requirement against that.

markstory replied:
I wasn't sure why we needed this interval class. I couldn't figure that out from the tests/code.

markstory replied:
It seems to duplicate some of the behavior in ChronosInterval. If folks need the additional precision beyond what PHP native objects provide they could use that package.

jimbo2150 replied:
I moved the `\DateInterval`-related code out of the `IntervalExpression` to keep it separate, so it's more of a utility class. It takes the `\DateInterval` properties (y, m, d, h, etc) and maps them into the database-specific units (YEAR, MONTH, DAY, HOUR, etc).

For comparison, `ChronosInterval` is used to reverse a `\DateInterval` object back into a period specification string `P1Y10D5H`.

This class instead creates an array `[ 'YEAR' => 1, 'DAY' => 10, 'HOUR' => 5 ]` that is used by `IntervalExpression` to generate `INTERVAL '1-0' YEAR_MONTH + INTERVAL '10 05:00:00.00' DAY_MICROSECOND` or the more specific `INTERVAL 1 YEAR + INTERVAL 10 DAY + INTERVAL 5 HOUR`. It also does some checks such as ensuring there is at least one valid value specified (creating an object with `P0Y` would not produce any change in a date it was applied to) as well as preventing relative or special formats (`second tuesday of march`) as those do not produce usable properties and are not supported by most databases. Since the methods are specific to database I left it under the `Cake\Database` namespace.

jimbo2150 committed:
- Updated per review suggestions

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 commented on src/Database/DateInterval.php:
```diff
@@ -0,0 +1,158 @@
+     * @param \DateInterval $clone Copy values from another DateInterval object.
+     * @throws \Exception
+     */
+    public function __construct($interval_spec, ?\DateInterval $clone = null)
```
Corrected

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
Merge branch '4.next' into 4.1-orm-interval

jimbo2150 committed:
- Fix single/double quote usage lint error

markstory commented:
The changes in #14738 should cover making interval expression with less complexity than a new expression object would.

jimbo2150 commented:
> The changes in #14738 should cover making interval expression with less complexity than a new expression object would.

It does cover window expression date ranges but does not appear to cover regular date arithmetic (`SELECT dateField + INTERVAL ...` or date function equivalents in other databases). Since it leaves the user up to adding text for the actual range (`1 day`, `2 month`, etc), MySQL common syntax is expected; It may not work on SQL Server since it expects slightly different syntax.

Doesn't appear to be interest in the arithmetic portion; closing.
